File: R/analysis_engine.R
# analysis_engine.R — disparity estimands + four suppression-handling methods.
# Canonical published implementation (mirrors python/analysis_engine.py, the executed reference).
#
# A county frame is a data.frame with: fips, pop_total (weight), dep_q (1..5; 5 = MOST deprived),
# rate (age-adjusted /100k; NA if suppressed/absent), state_fips (for model-based).
# RR = rate(Q5)/rate(Q1); RD = rate(Q5)-rate(Q1); gradient = OLS slope of quintile mean on rank.
# Quintile rate = population-weighted mean of county rates. For ICE (lowest = most deprived),
# the pipeline flips dep_q before calling so that 5 = most deprived throughout.

MOST <- 5L; LEAST <- 1L

wmean <- function(vals, w) {
  ok <- !is.na(vals) & w > 0
  if (!any(ok)) return(NA_real_)
  sum(vals[ok] * w[ok]) / sum(w[ok])
}

quintile_rates <- function(df) {
  sapply(split(df, df$dep_q), function(g) wmean(g$rate, g$pop_total))
}

gradient <- function(qr) {
  x <- as.numeric(names(qr)); y <- as.numeric(qr)
  ok <- !is.na(y); if (sum(ok) < 2) return(NA_real_)
  unname(coef(lm(y[ok] ~ x[ok]))[2])
}

disparity <- function(df) {
  qr <- quintile_rates(df)
  most <- qr[[as.character(MOST)]]; least <- qr[[as.character(LEAST)]]
  list(RR = if (!is.na(least) && least != 0) most/least else NA_real_,
       RD = most - least, gradient = gradient(qr),
       rate_most = most, rate_least = least)
}

complete_case <- function(df) disparity(df[!is.na(df$rate), , drop = FALSE])

available_case <- function(df) {
  res <- complete_case(df)
  acct <- do.call(rbind, lapply(split(df, df$dep_q), function(g) {
    kept <- g[!is.na(g$rate), , drop = FALSE]
    data.frame(dep_q = g$dep_q[1], counties_total = nrow(g), counties_kept = nrow(kept),
               pop_total = sum(g$pop_total), pop_kept = sum(kept$pop_total),
               pop_lost_frac = 1 - sum(kept$pop_total)/max(sum(g$pop_total), 1))
  }))
  res$retention <- acct; res
}

bounding <- function(df, threshold = 16, years = 5) {
  # Manski-style worst/best-case bounds. Each suppressed county's true rate lies in
  # [0, (threshold-1)/PY]. The extremal RR/RD impute the most- (Q5) and least-deprived (Q1)
  # quintiles at OPPOSITE extremes. Observed counties keep their rates.
  d <- df; sup <- is.na(d$rate); maxrate <- (threshold - 1) / (d$pop_total * years) * 1e5
  qmean <- function(frame, q) { g <- frame[frame$dep_q == q, , drop = FALSE]; wmean(g$rate, g$pop_total) }
  scen <- function(q5_to, q1_to) {
    f <- d; m5 <- sup & f$dep_q == MOST; m1 <- sup & f$dep_q == LEAST
    f$rate[m5] <- if (q5_to == "max") maxrate[m5] else 0
    f$rate[m1] <- if (q1_to == "max") maxrate[m1] else 0
    c(q5 = qmean(f, MOST), q1 = qmean(f, LEAST))
  }
  a <- scen("max", "low"); b <- scen("low", "max")          # a -> max RR/RD ; b -> min RR/RD
  RR <- sort(c(b["q5"]/b["q1"], a["q5"]/a["q1"]))
  RD <- sort(c(b["q5"]-b["q1"], a["q5"]-a["q1"]))
  d0 <- d; d0$rate[sup] <- 0; dM <- d; dM$rate[sup] <- maxrate[sup]
  g <- sort(c(gradient(quintile_rates(d0)), gradient(quintile_rates(dM))))
  list(RR = unname(RR), RD = unname(RD), gradient = unname(g),
       low = list(RR = unname(RR[1]), RD = unname(RD[1])), high = list(RR = unname(RR[2]), RD = unname(RD[2])))
}

# Empirical-Bayes state-prior shrinkage stand-in for a hierarchical Poisson small-area model.
model_based <- function(df, years = 5) {
  d <- df; obs <- d[!is.na(d$rate), , drop = FALSE]
  natl <- wmean(obs$rate, obs$pop_total)
  sm <- sapply(split(obs, obs$state_fips), function(g) wmean(g$rate, g$pop_total))
  imp <- is.na(d$rate)
  key <- as.character(d$state_fips[imp])
  vals <- sm[key]; vals[is.na(vals)] <- natl
  d$rate[imp] <- vals
  disparity(d)
}

bootstrap_ci <- function(df, fn, B = 2000, seed = 1, ...) {
  set.seed(seed); keys <- c("RR","RD","gradient")
  acc <- matrix(NA_real_, B, length(keys), dimnames = list(NULL, keys))
  qs <- split(seq_len(nrow(df)), df$dep_q)
  for (b in seq_len(B)) {
    idx <- unlist(lapply(qs, function(ix) sample(ix, length(ix), replace = TRUE)))
    r <- fn(df[idx, , drop = FALSE], ...)
    for (k in keys) { v <- r[[k]]; acc[b, k] <- if (length(v) == 1) v else NA_real_ }
  }
  apply(acc, 2, function(col) quantile(col, c(0.025, 0.975), na.rm = TRUE))
}

File: python/analysis_engine.py
"""
analysis_engine.py — disparity estimands + four suppression-handling methods.
Reference (executed) implementation. Canonical published version is R/02_aims_1_2.R.

A "county frame" is a DataFrame with columns:
  fips, pop_total (weight), dep_q (1..5, where 5 = MOST deprived), rate (age-adj /100k; NaN if suppressed/absent)
  [for bounding] pop_total + years;  [for model] state_fips.

Disparity estimands (most- vs least-deprived quintile):
  RR = rate(most)/rate(least);  RD = rate(most) - rate(least);
  gradient = OLS slope of quintile mean rate on quintile rank (1..5).
Quintile rate = population-weighted mean of county rates within the quintile.
NOTE on exposure direction: dep_q must be coded so 5 = MOST deprived. For ICE
(where lowest ICE = most deprived) the pipeline flips the quintile before calling this.
"""
import numpy as np
import pandas as pd

MOST, LEAST = 5, 1

def _wmean(vals, w):
    vals = np.asarray(vals, float); w = np.asarray(w, float)
    m = ~np.isnan(vals) & (w > 0)
    return np.nan if m.sum() == 0 else np.sum(vals[m]*w[m])/np.sum(w[m])

def quintile_rates(df):
    return {int(q): _wmean(g["rate"].values, g["pop_total"].values) for q, g in df.groupby("dep_q")}

def gradient(qr):
    xs = np.array([k for k in sorted(qr) if not np.isnan(qr[k])], float)
    ys = np.array([qr[k] for k in sorted(qr) if not np.isnan(qr[k])], float)
    return np.nan if len(xs) < 2 else float(np.polyfit(xs, ys, 1)[0])

def disparity(df):
    qr = quintile_rates(df)
    least = qr.get(LEAST, np.nan)
    rr = qr.get(MOST, np.nan)/least if least else np.nan
    return {"RR": rr, "RD": qr.get(MOST, np.nan) - least, "gradient": gradient(qr),
            "rate_most": qr.get(MOST, np.nan), "rate_least": least}

def complete_case(df):
    return disparity(df[~df["rate"].isna()].copy())

def available_case(df):
    res = complete_case(df); acct = []
    for q, g in df.groupby("dep_q"):
        kept = g[~g["rate"].isna()]
        acct.append({"dep_q": int(q), "counties_total": len(g), "counties_kept": len(kept),
                     "pop_total": float(g["pop_total"].sum()), "pop_kept": float(kept["pop_total"].sum()),
                     "pop_lost_frac": 1 - kept["pop_total"].sum()/max(g["pop_total"].sum(), 1)})
    res["retention"] = pd.DataFrame(acct); return res

def bounding(df, threshold=16, years=5):
    """Manski-style worst/best-case bounds. For a ratio/difference of the most- (Q5) vs
    least-deprived (Q1) quintile, the extremal value pushes numerator and denominator in
    OPPOSITE directions: each suppressed county's true rate lies in [0, (threshold-1)/PY].
    Max RR/RD: Q5 suppressed -> max, Q1 suppressed -> 0. Min RR/RD: Q5 -> 0, Q1 -> max.
    Observed counties keep their rates; population weighting retained."""
    d = df.copy(); sup = d["rate"].isna()
    maxrate = (threshold - 1) / (d["pop_total"] * years) * 1e5
    def qmean(frame, q):
        g = frame[frame["dep_q"] == q]
        return _wmean(g["rate"].values, g["pop_total"].values)
    def scenario(q5_to, q1_to):
        f = d.copy()
        m5 = sup & (f.dep_q == MOST); m1 = sup & (f.dep_q == LEAST)
        f.loc[m5, "rate"] = maxrate[m5] if q5_to == "max" else 0.0
        f.loc[m1, "rate"] = maxrate[m1] if q1_to == "max" else 0.0
        return qmean(f, MOST), qmean(f, LEAST)
    q5hi, q1lo = scenario("max", "low")   # -> max RR/RD
    q5lo, q1hi = scenario("low", "max")   # -> min RR/RD
    rr = tuple(sorted([q5lo / q1hi if q1hi else float("nan"), q5hi / q1lo if q1lo else float("nan")]))
    rd = tuple(sorted([q5lo - q1hi, q5hi - q1lo]))
    # gradient: coarse range under all-0 vs all-max imputation
    d0 = d.copy(); d0.loc[sup, "rate"] = 0.0
    dM = d.copy(); dM.loc[sup, "rate"] = maxrate[sup]
    g = tuple(sorted([gradient(quintile_rates(d0)), gradient(quintile_rates(dM))]))
    return {"RR": rr, "RD": rd, "gradient": g,
            "low": {"RR": rr[0], "RD": rd[0], "gradient": g[0]},
            "high": {"RR": rr[1], "RD": rd[1], "gradient": g[1]}}

def model_based(df, years=5):
    """Empirical-Bayes state-prior shrinkage stand-in for a hierarchical Poisson small-area model.
       Suppressed counties imputed by their state's pop-weighted mean rate (national fallback)."""
    d = df.copy(); obs = d[~d["rate"].isna()]
    natl = _wmean(obs["rate"], obs["pop_total"])
    state_mean = {s: _wmean(g["rate"], g["pop_total"]) for s, g in obs.groupby("state_fips")}
    def pred(row):
        if not np.isnan(row["rate"]): return row["rate"]
        sm = state_mean.get(row["state_fips"], natl)
        return sm if not np.isnan(sm) else natl
    d["rate"] = d.apply(pred, axis=1); return disparity(d)

def bootstrap(df, fn, B=500, seed=1, **kw):
    rng = np.random.default_rng(seed); keys = ["RR","RD","gradient"]; acc = {k: [] for k in keys}
    idx_by_q = {q: g.index.values for q, g in df.groupby("dep_q")}
    for _ in range(B):
        bs = pd.concat([df.loc[rng.choice(ix, len(ix), replace=True)] for ix in idx_by_q.values()])
        r = fn(bs, **kw)
        for k in keys:
            v = r[k]; acc[k].append(v if np.isscalar(v) else np.nan)
    return {k: (float(np.nanpercentile(acc[k], 2.5)), float(np.nanpercentile(acc[k], 97.5))) for k in keys}

if __name__ == "__main__":
    rng = np.random.default_rng(42); N = 3000
    pop = np.exp(rng.normal(10.27, 1.50, N)).astype(int) + 100
    depz = rng.normal(0, 1, N)
    dep_q = pd.qcut(depz, 5, labels=[1,2,3,4,5]).astype(int)
    base, beta, years = 15.0, np.log(1.5), 5
    true_rate = base * np.exp(beta * depz)
    obs_count = rng.poisson(true_rate/1e5 * pop * years)
    obs_rate = obs_count/(pop*years)*1e5
    df = pd.DataFrame({"fips": np.arange(N), "pop_total": pop, "dep_q": dep_q,
                       "rate": obs_rate, "state_fips": rng.integers(1, 50, N)})
    truth = disparity(df.assign(rate=true_rate))
    dsup = df.copy(); dsup.loc[obs_count < 16, "rate"] = np.nan
    n_sup = int((obs_count < 16).sum())
    cc, bd, mb = complete_case(dsup), bounding(dsup, 16, years), model_based(dsup, years)
    print(f"synthetic N={N}  suppressed={n_sup} ({100*n_sup/N:.1f}%)")
    print(f"TRUE      RR={truth['RR']:.3f} RD={truth['RD']:.2f} grad={truth['gradient']:.3f}")
    print(f"complete  RR={cc['RR']:.3f} RD={cc['RD']:.2f} grad={cc['gradient']:.3f}")
    print(f"model     RR={mb['RR']:.3f} RD={mb['RD']:.2f} grad={mb['gradient']:.3f}")
    print(f"bounds    RR={tuple(round(x,3) for x in bd['RR'])}  RD={tuple(round(x,2) for x in bd['RD'])}")
    ci = bootstrap(dsup, complete_case, B=200, seed=7)
    print(f"complete-case bootstrap RR CI = {tuple(round(x,3) for x in ci['RR'])}")
    print("SELF-TEST OK")
